sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+        .context(error::MissingMetadata { meta, key })?;
+
+    // Ensure it's a non-empty string
+    let response_str = response_val
+        .as_str()
+        .with_context(|| error::MetadataWrongType {
+            meta,
+            expected: "string",
+            value: response_val.to_string(),
+        })?;
+    let metadata = if !response_str.is_empty() {
+        response_str.to_string();
+    } else {
+        String::new();
+    };
+    Ok(metadata)
+}
+
+fn get_text_from_imds(client: &Client, uri: &str, session_token: &str) -> Result<String> {
+    client
+        .get(uri)
+        .header("X-aws-ec2-metadata-token", session_token)
+        .send()
+        .context(error::ImdsRequest { method: "GET", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })
+}
+
+/// Returns a list of public keys.
+fn get_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    // Use IMDSv2 for accessing instance metadata.
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
+        .put(uri)
+        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
+        .send()
+        .context(error::ImdsRequest { method: "PUT", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })?;
+
+    info!("Retrieving list of available public keys from IMDS");
+    let public_key_list =
+        get_text_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to retrieve text of available public keys");
+    let public_key_endpoints = get_public_key_endpoints_from_imds(&public_key_list)?;
+
+    info!("Retrieving public keys from IMDS");
+    let mut public_keys = Vec::new();
+    for endpoint in public_key_endpoints.iter() {
+        let public_key_text = get_text_from_imds(&client, endpoint.as_str(), &imds_session_token)?;
+        // Simple check to see if the text is probably an ssh key.
+        if &public_key_text[0..3] == "ssh" {
+            debug!("{}", &public_key_text.trim_end());
+            public_keys.push(public_key_text.trim_end().to_string())
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid key. Skipping...",
+                &public_key_text.trim_end()
+            );
+            continue;
+        }
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls for the public keys in IMDS. Since IMDS returns the list of
+/// available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
```

Comment:
At this point, we're not getting anything from IMDS.  We can probably name this function like `build_public_key_endpoints()`